File: src/main/java/dragonbook/inter/stmt/While.java
package dragonbook.inter.stmt;

import dragonbook.inter.expr.Expr;
import dragonbook.symbols.Type;

public class While extends Stmt {
    public Expr condition;
    public Stmt stmt;

    public While(Expr condition, Stmt stmt) {
        if (condition.type != Type.BOOL) condition.error("boolean required in while");
        this.condition = condition;
        this.stmt = stmt;
    }

    @Override
    public String toString() {
        return "while (" + condition + ") " + stmt;
    }
}

File: src/main/kotlin/dragonbook/inter/stmt/While.kt
package dragonbook.inter.stmt

import dragonbook.error.ParseError
import dragonbook.inter.expr.Expr
import dragonbook.symbols.Type

data class While(val condition: Expr, val stmt: Stmt) : Stmt {

    init {
        validate()
    }

    private fun validate() {
        if (condition.type != Type.BOOL) throw ParseError("boolean required in while", condition)
    }

    override fun toString(): String = "while ($condition) { $stmt }"
}